### DISCUSSION TITLE
Best practices for porting DirectML

### DISCUSSION THREAD

**Discussion by Axiverse** [amerkoleci/Vortice.Windows#257] [Ideas]
I'm working on porting DirectML and am willing to contribute the work to the general library if you'd want it here. I have some questions on conventions for some of the struct classes. I'm sure more questions will come up as I continue the port, but these are the ones I have currently.

## stackalloc
When calling some bindings, you [copy into a stackalloc array](https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.Direct3D12/ID3D12CommandQueue.cs#L17) before calling. I'm wondering why this is as SharpGenTools do create bindings which take in an array.

## DML_BINDING_DESC
DirectML heavy uses [DML_BINDING_DESC](https://docs.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_binding_desc) for all its operator descriptions and similar structures which use a type enum and a void ptr for the payload. Is there a recommended way of porting this to C#?

As they don't add any direct value, I've opted to collapse this out in favor of using the specializations which would then marshal directly into DML_BINDING_DESC. What do you think of this approach.

I have a draft pull request with some implementation and a partial port of the HelloDirectML. The part which has been ported compiles and runs without throwing any exceptions.

https://github.com/amerkoleci/Vortice.Windows/pull/256

**Comment by amerkoleci** [comment]
> ## stackalloc
> When calling some bindings, you [copy into a stackalloc array](https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.Direct3D12/ID3D12CommandQueue.cs#L17) before calling. I'm wondering why this is as SharpGenTools do create bindings which take in an array.
--------  **I probably need to fork SharpGenTools to handle those goodies**
> ## DML_BINDING_DESC
> DirectML heavy uses [DML_BINDING_DESC](https://docs.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_binding_desc) for all its operator descriptions and similar structures which use a type enum and a void ptr for the payload. Is there a recommended way of porting this to C#?
-------- **I will try to see how to handle it**
> As they don't add any direct value, I've opted to collapse this out in favor of using the specializations which would then marshal directly into DML_BINDING_DESC. What do you think of this approach.
> 
> I have a draft pull request with some implementation and a partial port of the HelloDirectML. The part which has been ported compiles and runs without throwing any exceptions.
> 
> #256

Thanks

**Reply by Axiverse** [reply]
I've updated the alloc structure to be closer to some other examples in the code. Everything handles it's own allocation and deallocation now. There's a number of allocations which need to happen for every call, but these descriptions shouldn't be used often in calls - only for setting up graphs.

**Comment by Axiverse** [comment]
On the stackalloc question, I'm just wondering why you do it and if it's necessary for me to do it as well.

**Comment by Axiverse** [comment]
I found some missing things. There's a few structures around DirectML 1.1 graphs which need to be wrapped. Otherwise I'd like to get some documentation in as well. Put a fix for SharpGenTools to (finally?) fix documentation inclusion. We will see. I might go ahead and start generating the documentation anyhow. 😃
